File: examples/scripts/python/activity_metrics.py
#!/usr/bin/env python3
"""Model Health Python SDK — Retrieve biomechanical metrics for an activity.

Demonstrates ``activity_metrics`` (single-activity dashboard metrics).

Usage:
    activity_metrics.py [<api_key>]
"""

import sys
from datetime import date, timedelta

from docopt import docopt

from modelhealth import (
    ModelHealthService,
    ModelHealthError,
    MetricValueScalar,
    MetricValueBilateral,
)
from _prompts import pick_one, confirm
from _utils import load_api_key, save_file

# Activities created by the mobile app for internal use — exclude from lists.
_INTERNAL_ACTIVITY_NAMES = {"calibration", "neutral"}


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    print("Connecting to Model Health...")
    try:
        service = ModelHealthService(api_key=api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")

    # Session
    print("\nFetching sessions...")
    sessions = service.session_list()
    if not sessions:
        sys.exit(
            "No sessions found. Create a session using the Model Health mobile app first."
        )

    print(f"\n{len(sessions)} session(s):\n")
    session = pick_one(
        sessions,
        "Select session",
        lambda s: f"[session ID: {s.id}]  session name: {s.session_name or '(unnamed)'}  subject: {s.name or '(unnamed)'}",
    )

    # Activities
    print(f"\nFetching activities for session ID: {session.id}...")
    all_activities = service.activity_list(session)
    activities = [a for a in all_activities if a.name not in _INTERNAL_ACTIVITY_NAMES]
    if not activities:
        sys.exit("No activities found in this session.")

    print(f"\n{len(activities)} activity/activities:\n")
    activity = pick_one(
        activities,
        "Select activity",
        lambda a: f"{a.name or a.id}  [{a.status}]",
    )

    # Activity metrics
    activity_label = activity.name or activity.id
    print(f"\nFetching metrics for '{activity_label}'...")
    try:
        metrics = service.activity_metrics(activity.id)
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch activity metrics: {exc}")

    flat = _flatten_metrics(metrics)
    if not flat:
        print("  No metrics available for this activity.")
    else:
        print("\nActivity metrics:\n")
        for name, value in flat.items():
            print(f"  {name}: {_format_value(value)}")

        if confirm("\nSave metrics as JSON?", default=False):
            slug = activity_label.replace(" ", "_")
            path = save_file(f"{slug}_metrics.json", metrics.to_json().encode("utf-8"))
            print(f"  Saved {path}")

    print("\nDone.")


def _flatten_metrics(metrics):
    """Collapse the grouped metrics into a flat name -> value mapping.

    Groups are discarded and each metric appears exactly once; the first
    occurrence of a name wins.
    """
    flat = {}
    for group in metrics.groups:
        for metric in group.metrics:
            if metric.name not in flat:
                flat[metric.name] = metric.value
    return flat


def _format_value(value):
    if isinstance(value, MetricValueScalar):
        return str(value.value) if value.value is not None else "—"
    if isinstance(value, MetricValueBilateral):
        left = str(value.left) if value.left is not None else "—"
        right = str(value.right) if value.right is not None else "—"
        return f"L {left}  R {right}"
    return repr(value)


if __name__ == "__main__":
    args = docopt(__doc__)
    try:
        main(load_api_key(args["<api_key>"]))
    except ModelHealthError as exc:
        print(f"Error: {exc}")
        sys.exit(1)

File: examples/scripts/ts/activity_metrics.ts
/**
 * Model Health TypeScript SDK — retrieve biomechanical metrics.
 * Mirrors examples/python/activity_metrics.py.
 *
 * Demonstrates activityMetrics (single-activity dashboard metrics).
 *
 * Usage:
 *   npx tsx activity_metrics.ts [<api_key>]
 */

import { ModelHealthClient } from '@modelhealth/modelhealth';
import type { ActivityMetrics, MetricValue } from '@modelhealth/modelhealth';
import { activityMetricsToJson } from '@modelhealth/modelhealth';
import {
  loadApiKey, INTERNAL_ACTIVITY_NAMES,
  pickOne, confirm, saveFile, closePrompts,
} from './_shared.js';

async function connect(apiKey: string): Promise<ModelHealthClient> {
  console.log('Connecting to Model Health...');
  const client = new ModelHealthClient({ apiKey, autoInit: false });
  await client.init();
  return client;
}

async function pickSession(client: ModelHealthClient) {
  console.log('\nFetching sessions...');
  const sessions = await client.sessionList();
  if (!sessions.length) {
    console.error('No sessions found. Create a session using the Model Health mobile app first.');
    process.exit(1);
  }

  console.log(`\n${sessions.length} session(s):\n`);
  return pickOne(sessions, 'Select session', s => {
    const sn = s.sessionName || '(unnamed)';
    const sub = s.name || '(unnamed)';
    return `[session ID: ${s.id}]  session name: ${sn}  subject: ${sub}`;
  });
}

async function pickActivity(client: ModelHealthClient, session: Awaited<ReturnType<typeof pickSession>>) {
  console.log(`\nFetching activities for session ID: ${session.id}...`);
  const allActivities = await client.activityList(session.id);
  const activities = allActivities.filter(a => !INTERNAL_ACTIVITY_NAMES.has(a.name ?? ''));
  if (!activities.length) { console.error('No activities found in this session.'); process.exit(1); }

  console.log(`\n${activities.length} activity/activities:\n`);
  return pickOne(activities, 'Select activity', a => `${a.name ?? a.id}  [${a.status}]`);
}

async function showMetrics(client: ModelHealthClient, activity: Awaited<ReturnType<typeof pickActivity>>) {
  const activityLabel = activity.name ?? activity.id;
  console.log(`\nFetching metrics for '${activityLabel}'...`);
  const metrics = await client.activityMetrics(activity.id);

  const flat = flattenMetrics(metrics);
  if (!flat.size) {
    console.log('  No metrics available for this activity.');
    return;
  }

  console.log('\nActivity metrics:\n');
  for (const [name, value] of flat) {
    console.log(`  ${name}: ${formatValue(value)}`);
  }

  if (await confirm('\nSave metrics as JSON?', false)) {
    const slug = activityLabel.replace(/ /g, '_');
    const p = saveFile(`${slug}_metrics.json`, Buffer.from(activityMetricsToJson(metrics), 'utf-8'));
    console.log(`  Saved ${p}`);
  }
}

async function main() {
  const args = process.argv.slice(2);
  const client = await connect(loadApiKey(args[0]));
  const session = await pickSession(client);
  const activity = await pickActivity(client, session);
  await showMetrics(client, activity);

  console.log('\nDone.');
}

/**
 * Collapse the grouped metrics into a flat name -> value map.
 * Groups are discarded and each metric appears exactly once; the first
 * occurrence of a name wins.
 */
function flattenMetrics(metrics: ActivityMetrics): Map<string, MetricValue> {
  const flat = new Map<string, MetricValue>();
  for (const group of metrics.groups) {
    for (const metric of group.metrics) {
      if (!flat.has(metric.name)) flat.set(metric.name, metric.value);
    }
  }
  return flat;
}

function formatValue(v: MetricValue): string {
  if (v.type === 'scalar') return v.value != null ? String(v.value) : '—';
  const l = v.left != null ? String(v.left) : '—';
  const r = v.right != null ? String(v.right) : '—';
  return `L ${l}  R ${r}`;
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
